tests/UnitTests/CoreGraphics.drawing/DrawingTest.cpp

```diff
@@ -70,47 +70,68 @@ CFStringRef testing::DrawTest::CreateOutputFilename() {
-
+    // Generate output filename (generally, TestCase.TestName.png).
     woc::unique_cf<CFStringRef> originalFilename{ CreateOutputFilename() };
 
     woc::unique_cf<CFMutableStringRef> filename{ CFStringCreateMutableCopy(nullptr, 0, originalFilename.get()) };
 
     CFStringFindAndReplace(filename.get(), CFSTR("DISABLED_"), CFSTR(""), CFRange{ 0, CFStringGetLength(filename.get()) }, 0);
     CFStringFindAndReplace(filename.get(), CFSTR("/"), CFSTR("_"), CFRange{ 0, CFStringGetLength(filename.get()) }, 0);
 
-    // This is only populated if CFStringGetCStringPtr fails.
-    std::unique_ptr<char[]> owningFilenamePtr;
-
-    char* rawFilename = const_cast<char*>(CFStringGetCStringPtr(filename.get(), kCFStringEncodingUTF8));
-    size_t len = 0;
-
-    if (!rawFilename) {
-        CFRange filenameRange{ 0, CFStringGetLength(filename.get()) };
-        CFIndex requiredBufferLength = 0;
-        CFStringGetBytes(filename.get(), filenameRange, kCFStringEncodingUTF8, 0, FALSE, nullptr, 0, &requiredBufferLength);
-        owningFilenamePtr.reset(new char[requiredBufferLength]);
-        rawFilename = owningFilenamePtr.get();
-        CFStringGetBytes(filename.get(),
-                         filenameRange,
-                         kCFStringEncodingUTF8,
-                         0,
-                         FALSE,
-                         (UInt8*)rawFilename,
-                         requiredBufferLength,
-                         &requiredBufferLength);
-        len = requiredBufferLength;
-    } else {
-        len = strlen(rawFilename);
+    auto drawingConfig = DrawingTestConfig::Get();
+
+    // Write the context image to the output file.
+    woc::unique_cf<CFDataRef> actualImageData{ _CFDataCreatePNGFromCGImage(image.get()) };
+    ASSERT_NE(nullptr, actualImageData);
+
+    woc::unique_cf<CFStringRef> outputPath{
+        CFStringCreateWithFormat(nullptr, nullptr, CFSTR("%s/%@"), drawingConfig->GetOutputPath().c_str(), filename.get())
+    };
+    ASSERT_NE(nullptr, outputPath);
+    outputPath.reset(_CFStringCreateAbsolutePath(outputPath.get()));
+
+    ASSERT_TRUE(_WriteCFDataToFile(actualImageData.get(), outputPath.get()));
+
+    // For comparisons ...
+    if (drawingConfig->GetMode() == DrawingTestMode::Compare) {
+        // ... load the reference image (same name, different directory as the output image)
+        woc::unique_cf<CFStringRef> referenceFilename{
+            CFStringCreateWithFormat(nullptr, nullptr, CFSTR("%s/%@"), drawingConfig->GetComparisonPath().c_str(), filename.get())
+        };
+        referenceFilename.reset(_CFStringCreateAbsolutePath(referenceFilename.get()));
+
+        woc::unique_cf<CGImageRef> referenceImage{ _CGImageCreateFromPNGFile(referenceFilename.get()) };
+        ASSERT_NE(nullptr, referenceImage);
+
+        // And fire off a comparator.
+        PixelByPixelImageComparator comparator;
+        auto delta = comparator.CompareImages(referenceImage.get(), image.get());
```

Comment:
Illegal transform. That will cause a leak.

---
In reply to: [89016958](https://github.com/Microsoft/WinObjC/pull/1438#discussion_r89016958) [](ancestors = 89016958)